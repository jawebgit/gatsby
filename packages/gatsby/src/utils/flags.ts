// Does this experiment run for only builds
type executingCommand = "build" | "develop" | "all"

export interface IFlag {
  name: string
  env: string
  description: string
  command: executingCommand
  telemetryId: string
  // Heuristics for deciding if a flag is experimental:
  // - there are known bugs most people will encounter and that block being
  // able to use Gatsby normally
  // - very few people have tested the feature so we're not sure if we've
  // uncovered even common problems.
  //
  // Flags should start as experimental but once all serious known bugs are
  // resolved and ~50+ people have tested it, experimental should be set to
  // false.
  experimental: boolean
  includedFlags?: Array<string>
  umbrellaIssue?: string
  noCi?: boolean
}

const activeFlags: Array<IFlag> = [
  {
    name: `FAST_DEV`,
    env: `GATSBY_EXPERIMENTAL_FAST_DEV`,
    command: `develop`,
    telemetryId: `FastDev`,
    experimental: false,
    description: `Enable all experiments aimed at improving develop server start time`,
    includedFlags: [
      `DEV_SSR`,
      `QUERY_ON_DEMAND`,
      `LAZY_IMAGES`,
      `PRESERVE_FILE_DOWNLOAD_CACHE`,
      `PRESERVE_WEBPACK_CACHE`,
    ],
  },
  {
    name: `DEV_SSR`,
    env: `GATSBY_EXPERIMENTAL_DEV_SSR`,
    command: `develop`,
    telemetryId: `DevSsr`,
    experimental: false,
    description: `SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.`,
    umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/28138`,
  },
  {
    name: `QUERY_ON_DEMAND`,
    env: `GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND`,
    command: `develop`,
    telemetryId: `QueryOnDemand`,
    experimental: false,
    description: `Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.`,
    umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/27620`,
    noCi: true,
  },
  {
    name: `LAZY_IMAGES`,
    env: `GATSBY_EXPERIMENTAL_LAZY_IMAGES`,
    command: `develop`,
    telemetryId: `LazyImageProcessing`,
    experimental: true,
    description: `Don't process images during development until they're requested from the browser. Speeds starting the develop server. Requires gatsby-plugin-sharp@2.10.0 or above.`,
    umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/27603`,
  },
  {
    name: `PRESERVE_WEBPACK_CACHE`,
    env: `GATSBY_EXPERIMENTAL_PRESERVE_WEBPACK_CACHE`,
    command: `all`,
    telemetryId: `PreserveWebpackCache`,
    experimental: false,
    description: `Don't delete webpack's cache when changing gatsby-node.js & gatsby-config.js files.`,
    umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/28331`,
  },
  {
    name: `PRESERVE_FILE_DOWNLOAD_CACHE`,
    env: `GATSBY_EXPERIMENTAL_PRESERVE_FILE_DOWNLOAD_CACHE`,
    command: `all`,
    telemetryId: `PreserveFileDownloadCache`,
    experimental: false,
    description: `Don't delete the downloaded files cache when changing gatsby-node.js & gatsby-config.js files.`,
    umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/28331`,
  },
  {
    name: `FAST_REFRESH`,
    env: `GATSBY_FAST_REFRESH`,
    command: `develop`,
    telemetryId: `FastRefresh`,
    experimental: false,
    description: `Use React Fast Refresh instead of the legacy react-hot-loader for instantaneous feedback in your development server. Recommended for versions of React >= 17.0.`,
    umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/28390`,
  },
  {
    name: `PARALLEL_SOURCING`,
    env: `GATSBY_EXPERIMENTAL_PARALLEL_SOURCING`,
    command: `all`,
    telemetryId: `ParallelSourcing`,
    experimental: true,
    description: `Run all source plugins at the same time instead of serially. For sites with multiple source plugins, this can speedup sourcing and transforming considerably.`,
    umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/28336`,
  },
]

export default activeFlags
