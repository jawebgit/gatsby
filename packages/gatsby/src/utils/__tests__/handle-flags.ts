import handleFlags from "../handle-flags"
import _ from "lodash"
import { IFlag } from "./flags"

jest.mock(`gatsby-core-utils`, () => {
  return {
    isCI: (): boolean => true,
  }
})

describe(`handle flags`, () => {
  const activeFlags: Array<IFlag> = [
    {
      name: `FAST_DEV`,
      env: `GATSBY_EXPERIMENTAL_FAST_DEV`,
      command: `develop`,
      description: `Enable all experiments aimed at improving develop server start time`,
      includedFlags: [`DEV_SSR`, `QUERY_ON_DEMAND`],
    },
    {
      name: `DEV_SSR`,
      env: `GATSBY_EXPERIMENTAL_DEV_SSR`,
      command: `develop`,
      description: `SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.`,
      umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/28138`,
    },
    {
      name: `QUERY_ON_DEMAND`,
      env: `GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND`,
      command: `develop`,
      description: `Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.`,
      umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/27620`,
      noCi: true,
    },
    {
      name: `ONLY_BUILDS`,
      env: `GATSBY_EXPERIMENTAL_BUILD_ME_FASTER`,
      command: `build`,
      description: `test`,
      umbrellaIssue: `test`,
    },
    {
      name: `ALL_COMMANDS`,
      env: `GATSBY_EXPERIMENTAL_SOMETHING_COOL`,
      command: `all`,
      description: `test`,
      umbrellaIssue: `test`,
    },
  ]

  const configFlags = {
    FAST_DEV: true,
    ALL_COMMANDS: true,
  }
  const configFlagsWithFalse = {
    ALL_COMMANDS: true,
    DEV_SSR: false,
  }
  const configWithFlagsNoCi = {
    QUERY_ON_DEMAND: true,
    DEV_SSR: true,
  }

  it(`returns validConfigFlags and a message`, () => {
    expect(handleFlags(activeFlags, configFlags, `develop`)).toMatchSnapshot()
  })

  it(`filters out flags marked false`, () => {
    expect(
      handleFlags(activeFlags, configFlagsWithFalse, `develop`)
        .enabledConfigFlags
    ).toHaveLength(1)
  })

  it(`filters out flags that are marked as not available on CI`, () => {
    expect(
      handleFlags(activeFlags, configWithFlagsNoCi, `develop`)
        .enabledConfigFlags
    ).toHaveLength(1)
  })

  it(`filters out flags that aren't for the current command`, () => {
    expect(
      handleFlags(activeFlags, configFlags, `build`).enabledConfigFlags
    ).toHaveLength(1)
  })

  it(`returns a message about unknown flags in the config`, () => {
    const unknownConfigFlags = handleFlags(
      activeFlags,
      { ALL_COMMANDS: true, FASTLY_DEV: true, SUPER_COOL_FLAG: true },
      `develop`
    )
    expect(unknownConfigFlags).toMatchSnapshot()
  })
})
