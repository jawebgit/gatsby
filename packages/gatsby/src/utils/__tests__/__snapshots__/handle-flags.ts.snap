// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle flags returns a message about unknown flags in the config 1`] = `
Object {
  "enabledConfigFlags": Array [
    Object {
      "command": "all",
      "description": "test",
      "env": "GATSBY_EXPERIMENTAL_SOMETHING_COOL",
      "name": "ALL_COMMANDS",
      "umbrellaIssue": "test",
    },
  ],
  "message": "The following flags are active:
- ALL_COMMANDS · (Umbrella Issue (test)) · test
",
  "unknownFlagMessage": "The following flag(s) found in your gatsby-config.js are not known:
- FASTLY_DEV (did you mean: FAST_DEV)
- SUPER_COOL_FLAG",
}
`;

exports[`handle flags returns validConfigFlags and a message 1`] = `
Object {
  "enabledConfigFlags": Array [
    Object {
      "command": "develop",
      "description": "Enable all experiments aimed at improving develop server start time",
      "env": "GATSBY_EXPERIMENTAL_FAST_DEV",
      "includedFlags": Array [
        "DEV_SSR",
        "QUERY_ON_DEMAND",
      ],
      "name": "FAST_DEV",
    },
    Object {
      "command": "all",
      "description": "test",
      "env": "GATSBY_EXPERIMENTAL_SOMETHING_COOL",
      "name": "ALL_COMMANDS",
      "umbrellaIssue": "test",
    },
    Object {
      "command": "develop",
      "description": "SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.",
      "env": "GATSBY_EXPERIMENTAL_DEV_SSR",
      "name": "DEV_SSR",
      "umbrellaIssue": "https://github.com/gatsbyjs/gatsby/discussions/28138",
    },
    Object {
      "command": "develop",
      "description": "Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.",
      "env": "GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND",
      "name": "QUERY_ON_DEMAND",
      "noCi": true,
      "umbrellaIssue": "https://github.com/gatsbyjs/gatsby/discussions/27620",
    },
  ],
  "message": "The following flags are active:
- FAST_DEV · Enable all experiments aimed at improving develop server start time
- ALL_COMMANDS · (Umbrella Issue (test)) · test
- DEV_SSR · (Umbrella Issue (https://github.com/gatsbyjs/gatsby/discussions/28138)) · SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.
- QUERY_ON_DEMAND · (Umbrella Issue (https://github.com/gatsbyjs/gatsby/discussions/27620)) · Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.

There are 2 other flags available that you might be interested in:
- ONLY_BUILDS · (Umbrella Issue (test)) · test
- YET_ANOTHER · (Umbrella Issue (test)) · test
",
  "unknownFlagMessage": "",
}
`;
